File: week-7/bkryusim/택배_배달과_수거하기.py
import math


def solution(cap, n, deliveries, pickups):
    answer = 0
    unload = 0
    load = 0

    for home in range(n - 1, -1, -1):
        # 이번집 배송하고 채울 물품 양
        unload -= deliveries[home]
        # 이번집 싣고 채울 빈박스 양
        load -= pickups[home]
        # 이 집에 왕복한 횟수
        count = max(math.ceil(-unload / cap), math.ceil(-load / cap))

        # 왕복한 만큼 물건 채워
        unload += cap * count
        load += cap * count

        answer += count * (home + 1) * 2

    return answer

File: week-7/protoseo/택배 배달과 수거하기.java
class Solution {
    public long solution(int cap, int n, int[] deliveries, int[] pickups) {
        int dIdx = -1;
        int pIdx = -1;
        for (int i = 0; i < n; i++) {
            if (deliveries[i] > 0) {
                dIdx = i;
            }
            if (pickups[i] > 0) {
                pIdx = i;
            }
        }
        long answer = 0;
        while (dIdx >= 0 || pIdx >= 0) {
            answer += ((Math.max(dIdx, pIdx) + 1) * 2);
            int temp = 0;
            while (dIdx >= 0 && temp < cap) {
                if (deliveries[dIdx] > 0) {
                    deliveries[dIdx]--;
                    temp++;
                }
                while (dIdx >= 0 && deliveries[dIdx] == 0) {
                    dIdx--;
                }
            }

            temp = 0;
            while (pIdx >= 0 && temp < cap) {
                if (pickups[pIdx] > 0) {
                    pickups[pIdx]--;
                    temp++;
                }
                while (pIdx >= 0 && pickups[pIdx] == 0) {
                    pIdx--;
                }
            }
        }
        return answer;
    }
}
